src/pin.js

```diff
@@ -0,0 +1,158 @@
+/* eslint max-nested-callbacks: ["error", 8] */
+
+'use strict'
+
+const expect = require('chai').expect
+const fs = require('fs')
+const path = require('path')
+
+const testfile = fs.readFileSync(path.join(__dirname, './data/testfile.txt'))
+
+module.exports = (common) => {
+  describe('.pin', () => {
+    let ipfs
+
+    before(function (done) {
+      // CI takes longer to instantiate the daemon,
+      // so we need to increase the timeout for the
+      // before step
+      this.timeout(20 * 1000)
+
+      common.setup((err, factory) => {
+        expect(err).to.not.exist
+        factory.spawnNode((err, node) => {
+          expect(err).to.not.exist
+          ipfs = node
+          populate()
+        })
+      })
+
+      function populate () {
+        const expectedMultihash = 'Qma4hjFTnCasJ8PVp3mZbZK5g2vGDT4LByLJ7m8ciyRFZP'
+
+        ipfs.files.add(testfile, (err, res) => {
+          expect(err).to.not.exist
+
+          expect(res).to.have.length(1)
+          expect(res[0].hash).to.equal(expectedMultihash)
+          expect(res[0].path).to.equal(expectedMultihash)
+          done()
+        })
+      }
+    })
+
+    after((done) => {
+      common.teardown(done)
+    })
+
+    describe('callback API', () => {
+      // 1st, because ipfs.files.add pins automatically
+      it('.rm', (done) => {
+        const hash = 'Qma4hjFTnCasJ8PVp3mZbZK5g2vGDT4LByLJ7m8ciyRFZP'
+
+        ipfs.pin.rm(hash, { recursive: true }, (err, pinset) => {
+          expect(err).to.not.exist
+          expect(pinset).to.exist
+          ipfs.pin.ls({ type: 'direct' }, (err, pinset) => {
+            expect(err).to.not.exist
+            expect(pinset).to.be.empty
+            done()
+          })
+        })
+      })
+
+      it('.add', (done) => {
+        const hash = 'Qma4hjFTnCasJ8PVp3mZbZK5g2vGDT4LByLJ7m8ciyRFZP'
+
+        ipfs.pin.add(hash, { recursive: false }, (err, pinset) => {
+          expect(err).to.not.exist
+          expect(pinset[0]).to.be.equal(hash)
+          done()
+        })
+      })
+
+      it('.ls', (done) => {
+        ipfs.pin.ls((err, pinset) => {
+          expect(err).to.not.exist
+          expect(Object.keys(pinset).length > 0).to.equal(true)
+          done()
+        })
+      })
+
+      it('.ls type direct', (done) => {
+        ipfs.pin.ls({ type: 'direct' }, (err, pinset) => {
+          expect(err).to.not.exist
+          expect(Object.keys(pinset).length > 0).to.equal(true)
+          done()
+        })
+      })
+
+      it('.ls type indirect', (done) => {
+        ipfs.pin.ls({ type: 'indirect' }, (err, pinset) => {
+          expect(err).to.not.exist
+          expect(Object.keys(pinset).length > 0).to.equal(true)
+          done()
+        })
+      })
+
+      it('.ls type recursive', (done) => {
+        ipfs.pin.ls({ type: 'recursive' }, (err, pinset) => {
+          expect(err).to.not.exist
+          expect(Object.keys(pinset).length > 0).to.equal(true)
+          done()
+        })
+      })
```

Comment:
These three tests above could be simplified to

function testLs (type) {
  return it(`.ls type ${type}`, (done) => {
  })
}